src/components/combo/combo.ts

```diff
@@ -0,0 +1,654 @@
+  @watch('pipeline')
+  protected async pipeline() {
+    this.dataState = await this.dataController.apply([...this.data]);
+    this.navigationController.active = 0;
+  }
+
+  @watch('open')
+  protected toggleDirectiveChange() {
+    if (!this.target) return;
+    this.toggleController.target = this.target;
+    this.target.setAttribute('aria-expanded', this.open ? 'true' : 'false');
+  }
+
+  constructor() {
+    super();
+
+    this.toggleController = new IgcToggleController(this, {
+      target: this.target,
+      closeCallback: async () => {
+        if (!this.handleClosing()) return;
+        this.open = false;
+
+        await this.updateComplete;
+        this.emitEvent('igcClosed');
+      },
+    });
+
+    this.addEventListener('focus', () => {
+      this.emitEvent('igcFocus');
+    });
+
+    this.addEventListener('blur', () => {
+      this.emitEvent('igcBlur');
+    });
+
+    this.addEventListener(
+      'keydown',
+      this.navigationController.navigateHost.bind(this.navigationController)
+    );
+  }
+
+  protected override async firstUpdated() {
+    await this.updateComplete;
+    this.requestUpdate();
+  }
+
+  protected override async getUpdateComplete() {
+    const result = await super.getUpdateComplete();
+    await this.toggleController.rendered;
+    return result;
+  }
+
+  /**
+   * Returns the current selection as a list of commma separated values,
+   * represented by the display key, when provided.
+   */
+  public get value() {
+    return this.selectionController.getValue(
+      Array.from(this.selectionController.selected)
+    );
+  }
+
+  @watch('value')
+  protected validate() {
+    this.updateComplete.then(() => this.reportValidity());
+  }
+
+  /** Checks the validity of the control. */
+  public reportValidity() {
+    this.invalid = this.required && !this.value;
+    return !this.invalid;
+  }
+
+  /** A wrapper around the reportValidity method to comply with the native input API. */
+  public checkValidity() {
+    return this.reportValidity();
+  }
+
+  /** Sets focus on the component. */
+  public override focus(options?: FocusOptions) {
+    this.target.focus(options);
+  }
+
+  /** Removes focus from the component. */
+  public override blur() {
+    this.target.blur();
+  }
+
+  /**
+   * Selects the options in the list by either value or key value.
+   * If not argument is provided all items will be selected.
+   * @param { T[] | Values<T>[] } items - A list of values or values as set by the valueKey.
+   */
+  public select(items?: T[] | Values<T>[], emit = false) {
+    this.selectionController.select(items, emit);
+  }
+
+  /**
+   * Deselects the options in the list by either value or key value.
+   * If not argument is provided all items will be deselected.
+   * @param { T[] | Values<T>[] } items - A list of values or values as set by the valueKey.
+   */
+  public deselect(items?: T[] | Values<T>[], emit = false) {
+    this.selectionController.deselect(items, emit);
```

Comment:
I prefer to skip the option for emitting an event.
It's an API call

Reply:
Done.